src/v/kafka/server/connection_context.cc

```diff
@@ -394,11 +398,12 @@ connection_context::dispatch_method_once(request_header hdr, size_t size) {
                        */
                       if (d.failed()) {
                           return f.discard_result()
-                            .handle_exception([](std::exception_ptr e) {
+                            .handle_exception([&h_probe](std::exception_ptr e) {
                                 vlog(
                                   klog.info,
                                   "Discarding second stage failure {}",
                                   e);
+                                h_probe.request_errored();
```

Comment:
I'd be tempted to put this in the finally block to colocate it with `self->_server.probe().service_error();`

I see that `self->_server.probe()` is both completed and errored, which seems like it might double-count completions since that's also done in the destructor of request_tracker.

Should `h_probe::_requests_completed` include errored?

It looks to me that the use of `h_probe` in `maybe_process_responses`, is a double-counted error.

Prodding these probes is a bit disparate, it looks easy to miss one, especially under maintenance. I wonder if it can be cleaned up a bit with either an RAII object and/or moving it to a different part of the code?

Reply:
Moved it to the `finally` block.

 `h_probe::_requests_completed` is both successfully completed and errored. 

As for whether I'm double counting errors in `maybe_process_responses` I don't believe so. It's done in a catch block for a failed write to a connection and the error isn't propagate up anywhere else. So The only point to actually count that particular errored request is in that catch block.

Definitely agree that the prodding of the errored probe is disparate. Though I'm not sure if much can be done to help this as our handling of request errors is disparate. Take the previously describe error in `maybe_process_responses`, if I want to record that error in the probe I need to do so in the catch block as the error isn't propagated to any other point in the code. It's a similar story for other places errors occur on the request path. I don't believe RAII objects would help either here as it would introduce an extensive amount of explicit moves as I try to get the object into all the disparate error handlers.

Reply:
`self->_server.probe()` definitely seems to double count completions in the finally block as you have said. Will add an additional commit to fix this bug.

Reply:
I've moved as much as I can to the `request_tracker` let me know what you think. I could probably move sent/recv'd bytes to that class as well.